Corrects error throwing...

S0AndS0 opened the pull request:
... removes most JQuery dependencies, and adds privacy notice page.


___


For future reference there are four ways to throw Error objects that preserve the stack-trace and allow for proper catching.



1. throw a general error (_one-off_)...


```JavaScript
const fn_someThing = (arg) {
  if (arg.toLowerCase() !== 'some thing') {
    throw new Error('Argument was not "some thing"');
  }
}


try {
  fn_someThing();
} catch (e) {
  console.errror(e.message);
  throw e;
}
```


------



2. throw a custom error that's used once, as well as filter by `name`, while also including the stack trace...


```JavaScript
const fn_anotherThing = (arg) => {
  if (arg.toLowerCase() !== 'some thing') {
    throw {
      name: 'anotherThingError',
      message: 'anotherThing was not "another thing"',
      stack: new Error(this.message).stack
    };
  }
}


try {
  fn_anotherThing();
} catch (e) {
  if (e.name !== 'anotherThingError') {
    throw e;
  }
  console.error(`${e.name} reports -> ${e.message}`);
  // Perhaps `try` something else later, but for now re-throw
  throw e;
}
```


------


3. throw a custom error that's used more than once, and remain compatible with legacy JavaScript...


```JavaScript
function customError(message) {
  this.name = 'customError';
  this message = (message || '');
  this.stack = new Error(this.message).stack;
}
customError.prototype = Error.prototype;


const fn_throwAll = (arg) => {
  throw new customError('Custom error message');
}


try {
  fn_throwAll();
} catch (e) {
  if (e.name !== 'customError') {
    throw e;
  }
  console.error(e.message);
}
```


------


4. throw a custom error class and have TypeScript transpile instead...


```TypeScript
class Chucker extends Error {
  constructor(message: string = '') {
    /**
     * Note, `arguments` are a built-in keyword for **non-arrow** functions and
     *  class methods, which contain a list of **all** arguments passed
     *  including any named parameters too, eg. in this case `message` plus
     *  anything else that was passed to this constructor method.
     */
    super(...arguments);
    this.message = message;
    this.name = 'Chucker';
  }
}


const fn_chuckAll = () => {
  throw new Chucker('Throwing up the stack with class');
};


try {
  fn_chuckAll();
} catch (e) {
  if (e.name !== 'Chucker') {
    throw e;
  }
  console.error(`Classy error message from ${e.name} -> ${e.message}`);
  throw e;
}
```


## Attribution


- [StackOverflow -- How do I create a custom error in JavaScript?](https://stackoverflow.com/questions/783818)


___


## File-by-file break-down


**`public/index.html`**


- Removes image domain from meta-data address, browsers _should_ be able to figure out relative paths and this allows for easier testing on GitHub pages from a sub-directory

- Moves script `src` to be within the `head` element, note if errors about things being undefined pop then _wrapping_ things within page load callbacks may be required, eg...


```JavaScript
document.addEventListener('load', (_event) => {
  /*...*/
});
```


- Adds element for Google Translate scripts to populate with language `select` options, note current tests on GitHub Pages showed that scripts where loading but element was for some reason being ignored, thus previous translation code has not yet been removed


- Minor wording corrections to navigation list(s), however, tests showed that side-bar nav element was not being displayed even when page width was _squished_... is there a missing `@media` query within the CSS?


------


**`public/privacy.html`**


This file is a starting place for future expansion on how client data is handled and should be expended upon to include who clients should contact to have their data removed, corrected, and or downloaded.


------


**`ts/builder/builder.ts`**


- Corrects error throwing


------


**`ts/js/app.ts`**


- Moved from `public/js/app.js` and pulls in functions from `public/js/utils.js`

- Adds TypeScript hints and JSDoc comment blocks

> Note, most JSDoc comments will require further expansion prior to building complete API documentation, but that seems to be the case for most of the scripts at this time


------


**`ts/translate/translate_selection.ts`**


- Adds callback functions and hooks for injecting Google Translate features into HTML pages

> Note, TypeScript hints for `google.translate.TranslateElement` require a bit more research in order to remove _`any`_


------


**`ts/view/audio.ts`**


- Corrects error throwing



------


**`ts/view/braille_controller.ts`**


- Removes JQuery dependencies

- Corrects error throwing

- Adds more JSDoc comments

- Removes `getAllEvents` and adds replaces with`bindAllEvents` method



------


**`ts/view/view.ts`**


- Removes JQuery dependencies

> Note `querySelectorAll` stuff seems a little messy and it may be better to have a function/method for making code a bit terser

- The inner `for` loop within `createGrid` needs a second opinion, even before alterations the `combinedDataAndHeaders[0]` seems a bit off

- Adds `getElementsByRole` function for combining `getElementsByTagName` and filtering by `role` attribute

oferb commented on public/privacy.html:
```diff
@@ -0,0 +1,26 @@
+
+  <body>
+    <strong>
+      Currently the technologies utilized by this site may cause submitted graph data to be saved within server logs.
```
We don't have server logs :-)

S0AndS0 replied:
While that may be true at this time there is no way to verify it now or in the future publicly, and if this project is deployed elsewhere there is no way to prevent those server-admins from keeping logs.


While it's not a _water tight_ disclaimer, and really should be treated as a template for future expansion, not having some form of notice is going to attract the wrong kind of attention.

oferb replied:
So how about we write that in sensoryinterface.com we don't log it and in other websites it may happen?

S0AndS0 replied:
That's an option, however there's still a chance that logging can happen and there's still the possibility that data can be intercepted. So keeping the notice that private/confidential information shouldn't be transmitted is likely a good idea.

oferb replied:
I agree that it's good to write that in general, but, people who are not technical might understand it differently than you intend.

I think we should write something like:
"The sensoryinterface.com doesn't log or store user data, though technically it's possible to do so."

S0AndS0 replied:
There's nothing technical in the current notice, it simply requests that private or confidential data not be transmitted to this project.

Regardless of if the server logs anything currently or in the future, data can still be intercepted through no fault of hosting server configurations.

Until this project transitions to techniques that don't require current workarounds for data parsing, I think the best way to protect clients from abuse is set their expectation of privacy/security accordingly.

oferb replied:
> data can still be intercepted through no fault of hosting server configurations
How? We use HTTPS. You mean a MITM attack?

oferb replied:
Also, if you're dealing with confidential company data, there's usually company policies for not posting that data on other websites. Those same policies probably prohibit installing Chrome extensions.

So, practically speaking, even if the extension does not transmit data, the company won't approve it because it would need to get a security specialist to evaluate the Chrome extension's code to whitelist it. Also, the code can change at any time, so that doesn't actually help. They would need to copy the extension into a private company extension registry. It's easier to host the files locally in that case.

So in practice, I think companies that care about confidential data won't put that data in someone else's website regardless of the privacy notice that website put out.

If you're talking about personal confidential data, then it's mostly a trust thing, because websites can send data across the internet regardless of the url changing.

Bottom line, we're not interested in user data and we don't log it, but you need to trust us for that. I think that's what this privacy policy should state. Technically speaking, we're currently using Firestore Hosting so AFAIK I don't have access to server logs (Google does though, but they have a policy where they probably can't use that for anything except things like debugging issues).

For people very concerned with privacy, we can provide instructions for running the server locally (as we do in the contrib instructions). So someone can do that if they prefer. They would need git and something to host on localhost, say python. That would solve the issue I think as it's 100% local and it can be done with today's setup.

S0AndS0 replied:
HTTPS is not a barrier but a speed-bump, the _pad-lock_ in the address bar is very analogous to a physical pad-lock in this regard. There have been vulnerabilities disclosed in the past for protocol and there's likely to be more in the future.

I think it's best to not trust that companies have policies, enforce'em, or have the ability to detect information leaks. Regardless humans are prone to error and code can be a tool to expedite such errors.

Google is in the business of data collection, they've nothing but incentive to keep such logs, and already have the infrastructure to do so.

In the mean-time it's probably a good idea to divide the code features between what's used for language translations vs. application code.

oferb replied:
Ok, so if you're worried about privacy, you can use the local server setup, right?

> mean-time it's probably a good idea to divide the code features between what's used for language translations vs. application code

oferb replied:
I suggest we move the `privacy.html` file to a different PR and continue the discussion there, to not block merging this PR.

S0AndS0 replied:
Indeed and not entirely, there are still calls out to third-parties for JavaScript and CSS code that isn't within this repository.

I mean that code translating tutorials and other pages should be organized within a separate sub-directory from the code that preforms braille/audio translations.

Just because they state that it's not used for adverts does not mean that client data is not otherwise stored and/or process either by Google, or it's partners, or those that force Google to share such data.

The "cloud" really means "someone-else's computer". You only own the data that is stored on infrastructure directly under your control, and even that ownership is tenuous and may not last. I don't host _private_ repositories on GitHub, nor do I use VPNs for _privacy_ for the same reasons; I don't have faith in claims of companies that cannot be verified.

I'll do a `git rm --cache public/privacy.html` and once this PR is closed I'll be opening one specifically for it as requested.

S0AndS0 commented:
> @oferb
> We'd like to keep manual translations where available. Does the code currently do that?

Current changes do not seem to effect manual translations, and in future PRs (after sorting out Google Translate display of options) I'll try adding an _ignore list_ that mitigates chances for conflicts between auto and manual translations.

oferb commented:
Thinking about it a little more, I'd like to suggest that we don't add dynamic translations at this point in the project. That's not our bottleneck and I prefer to keep things simple as much as possible and focus on the bottlenecks first.

S0AndS0 commented:
I've removed the Google Translate features as suggested.


Also I noticed that some JavaScript files didn't have a TypeScript source, so I've moved/renamed'em and corrected the resulting build errors too.

mohammad-suliman commented on ts/view/view.ts:
```diff
@@ -43,41 +47,66 @@ function processData() {
-    $(element).on('touchmove', onCellChange);
-    $(element).on('touchleave', stopSoundPlayback);
-    $(element).on('touchcancel', stopSoundPlayback);
+  getElementsByRole('div', 'gridcell').forEach((element: HTMLElement, _index: number) => {
```
I think you don't need to use this function, rather, you can use selectors. See for example this answer on stackoverflow:
https://stackoverflow.com/questions/46648761/how-to-select-element-by-role-attribute-and-filter-by-class-with-vanilla-js

S0AndS0 replied:
Oh now that's superb. I'll be sure to add that to the attribution list too.

mohammad-suliman commented on ts/view/view.ts:
```diff
@@ -87,13 +116,13 @@ function onClick(event) {
 
 function addNavigationToGrid() {
-  $('div[role="gridcell"]').each(function (index, gridCell) {
+  getElementsByRole('div', 'gridcell').forEach((gridCell: HTMLElement, index: number) => {
```
Also here...

S0AndS0 replied:
Yep, next batch of commits will include the `document.querySelectorAll` trick for sure.

S0AndS0 commented on ts/view/view.ts:
```diff
@@ -43,41 +47,66 @@ function processData() {
+ * @param {string} role
+ * @returns {HTMLElement[]}
+ */
+const getElementsByRole = (element_name: string, role: string): HTMLElement[] => {
```
Heh, now that's a habit I'm still struggling to break

S0AndS0 commented on ts/view/view.ts:
```diff
@@ -43,41 +47,66 @@ function processData() {
+      gridCell.setAttribute('col', colIndex.toString());
       gridRow.append(gridCell);
     }
-    grid.append(gridRow);
```
I think so, otherwise how are rows going to be added to the grid?

S0AndS0 commented on ts/view/view.ts:
```diff
@@ -43,41 +47,66 @@ function processData() {
+    gridRow.setAttribute('class', 'row');
+
+    /**
+     * The `combinedDataAndHeaders[0]` feels like it should be `combinedDataAndHeaders[rowIndex]`
```
Okay thanks for confirming my beliefs. If it's not incorrect I'll not change it just yet though.

S0AndS0 commented on ts/view/braille_controller.ts:
```diff
@@ -47,6 +54,20 @@ class BrailleController {
+   */
+  static bindAllEvents(element: HTMLTextAreaElement, callback) {
+    let events_list = [];
+    for (var key in this) {
```
Huh, not sure where the `var` was introduced... here I thought I had broken myself of such habits. Regardless it'll be corrected soon.

S0AndS0 commented on ts/view/braille_controller.ts:
```diff
@@ -7,33 +7,40 @@ class BrailleController {
+    const brailleControllerLabel: HTMLLabelElement = document.createElement('label');
+    brailleControllerLabel.innerHTML = 'Use Left / Right arrows to navigate the graph.<br> Use space bar to get more info about the value under the cursor.';
+    brailleControllerLabel.setAttribute('for', 'brailleControllerText');
     parent.appendChild(brailleControllerLabel[0]);
```
Indeed, I'll remove the indexes within the next batch of commits

S0AndS0 commented on ts/languages/i18n.ts:
```diff
@@ -0,0 +1,87 @@
+
+/**
+ * Save language in local storage if not already there, update HTML
+ * @param {string} selectLanguage
```
Wups, this too will be corrected in the next set of commits

S0AndS0 commented on ts/js/app.ts:
```diff
@@ -0,0 +1,66 @@
+const changeLanguage = (language: string) => {
+  document.getElementById('language-popup').setAttribute('modal', 'toggle');
+  // Note, TypeScript complained about `.modal` not being present on JQuery element
+  // $('#language-popup').modal('toggle');
```
Super, I'll remove'em in the next set of commits

S0AndS0 commented on ts/js/app.ts:
```diff
@@ -0,0 +1,66 @@
+ * @param {string} name
+ */
+const deleteQueryParam = (name: string) => {
+  const queryString = window.location.search;
```
Excellent catch!

I totally missed those and line `17`, next push will include commits that correct this.

S0AndS0 commented on ts/builder/builder.ts:
```diff
@@ -111,29 +111,33 @@ function onRadioChange(radio: HTMLInputElement) {
 /**
```
There may be a needed final `try`/`catch` to be added to prevent errors from stopping execution.


Instead of using a function to get an error's message, read the `.message` property from thrown error objects, eg...


```JavaScript
try {
  throw New Error('Uh-ohs');
} catch (e) {
  console.error(e.message);
}
```

... also `console.error()` can handle error objects without any property deconstruction...


```JavaScript
try {
  throw New Error('Uh-ohs');
} catch (e) {
  console.error(e);
}
```


... which not only is the _proper_ way of logging errors without breaking execution, it also preserves and presents the stack-trace for easier debugging.

mohammad-suliman replied:
Thanks! In our case, the errors are displayed for the user. Not sure if the code needs to handle this better.

S0AndS0 replied:
Here are two examples for capturing and parsing all errors, note there are subtitle differences for parameters that are passed depending of if adding an event listener vs. overwriting the `onerror` event...


```javascript
window.addEventListener('error', (message, source, lineno, colno, error) => {
  console.error('window.addEventListener caught error...');
  console.error(`message -> ${message}`);
  console.error(`source -> ${source}`);
  console.error(`lineno -> ${lineno}`);
  console.error(`colno -> ${colno}`);
  console.error('error...');
  console.error(error);
});


window.onerror = (msg, url, lineNo, columnNo, error) => {
  console.error('window.onerror caught error...');
  console.error(`msg -> ${msg}`);
  console.error(`url -> ${url}`);
  console.error(`lineNo -> ${lineNo}`);
  console.error(`columnNo -> ${columnNo}`);
  console.error('error...');
  console.error(error);
};


try {
  throw new Error('Uh-oh');
} catch (e) {
  console.error(e.message, 'spaghetti-ohs');
  throw e;
}
```


... which may be a useful place to put code for presenting errors to clients.


## Attribution

- [Mozilla Developer -- `GlobalEventHandlers.onerror`](https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onerror)

S0AndS0 commented on ts/builder/builder.ts:
```diff
@@ -192,35 +202,22 @@ function needToTranspose(data: string[][]): boolean {
+        break;
       }
     }
+    // Why is this boolean inverted?
```
Thanks, I'm adding a note in-place of that question then.

S0AndS0 commented on public/builder/index.html:
```diff
@@ -96,9 +96,10 @@ <h2>Options</h2>
     <br>
-    <a href="tutorial.html">Tutorial</a>
-    <a href="about.html">About</a>
+    <!-- Following require fixing to avoid 404 errors -->
```
The `href` attributes could be changed to `../tutorial.html` and `../about.html`, because those files exist in the parent directory, but these sorts of things will likely continue to pop-up until some site-builder tools are adopted.


Recently pushed commits should correct these links in the meantime.

oferb replied:
I think we'll go to the SPA route and use components. Something like this:

```
window.onload = () => {
  // Load components
  const headerLoading = new rxjs.Observable(observer => {
    $("#header").load("components/header.html", () => observer.next());
  });
  const contentLoading = new rxjs.Observable(observer => {
    $("#content").load("components/content.html", () => observer.next());
  });
  const footerLoading = new rxjs.Observable(observer => {
    $("#footer").load("components/footer.html", () => observer.next());
  });
  // Wait until all components are loaded
  rxjs.zip(
    headerLoading,
    contentLoading,
    footerLoading,
  ).subscribe(() => {
    // Components are loaded, display them
    $("#loading").css({ display: 'none' });
    $("#components").css({ display: 'block' });
  });
};
```

mohammad-suliman replied:
Thanks for the catch @S0AndS0! I missed that somehow 😊

S0AndS0 replied:
@oferb the more dynamic stuff that's loaded on the client the more likely it is that pages will not load properly; the benefit of either MustacheJS or Jekyll is that pages are compiled and served with minimal to no client-side scripts required.

- Jekyll has little to no overhead if a preexisting theme is used, and customizing can be incremental

- MustacheJS has quite a bit of stuff that would need to be written for this project, but that also means there's no theme second guessing style and markup decisions

- Either can make use of Scss for compiling CSS too; Jekyll uses Ruby based Sass and Node Sass could be added if MustacheJS was used instead

@mohammad-suliman welcome for sure; ya have caught quite a few of mine ;-)

oferb replied:
@S0AndS0 Thanks for the suggestion. As of now, we'll continue using vanilla js without those libraries, but we'll keep it in mind.
